File: 1. Java DS/BinaryHeap.java
import java.util.Arrays;

public class BinaryHeap {

    private Integer[] heap;
    private int n = 0;
    private int capacity = 8;

    public BinaryHeap() {
        this.heap = new Integer[capacity];
    }

    public void insert(Integer value) {
        if (n+1 == capacity) {
            capacity *= 2;
            heap = Arrays.copyOf(heap, capacity);
        }
        heap[++n] = value;
    }

    public Integer deleteMax() {
        Integer max = heap[1];
        swap( 1, n);
        heap[n--] = null;
        sink(1);
        return max;
    }

    private void swim(int k) {
        while (k/2 > 1 && heap[k/2] < heap[k]) {
            swap(k/2, k);
            k /= 2;
        }
    }

    private void sink(int k) {
        while (k*2 < n) {
            int j = k*2;
            if (heap[j] < heap[j+1]) {
                j++;
            }
            if (!(heap[k] < heap[j])) {
                break;
            }
            swap(k, j);
            k = j;
        }
    }

    private void swap(int i, int j) {
        int t = heap[i];
        heap[i] = heap[j];
        heap[j] = t;
    }
}

File: python_data_structures/binary_heap/binary_heap.py
class BinaryHeap:
    def __init__(self):
        self.heap = [0]
        self.n = 0

    def is_empty(self):
        return self.n == 0

    def insert(self, k):
        self.n += 1
        self.heap.append(k)
        self.__swim(self.n)

    def del_max(self):
        max_ = self.heap[1]
        self.__swap(1, self.n)
        self.heap[self.n] = None
        self.n -= 1
        self.__sink(1)
        return max_

    def __swim(self, k):
        while k > 1 and self.heap[k//2] < self.heap[k]:
            self.__swap(k//2, k)
            k //= 2

    def __sink(self, k):
        while 2*k < self.n:
            j = 2*k
            if self.heap[j] < self.heap[j+1]:
                j += 1
            if not self.heap[k] < self.heap[j]:
                break

            self.__swap(k, j)
            k = j

    def __swap(self, k, j):
        tmp = self.heap[k]
        self.heap[k] = self.heap[j]
        self.heap[j] = tmp

